batcher/sweep_batch.go

```diff
@@ -0,0 +1,1353 @@
+// timeout must not be too far away from the other timeouts of existing sweeps.
+// If this batch contains a single sweep that spends to a non-wallet address we
+// are not accepting other sweeps as batching is pointless in this case.
+func (b *Batch) AcceptsSweep(sweep *Sweep) bool {
+	b.sweepsLock.Lock()
+	defer b.sweepsLock.Unlock()
+
+	if sweep == nil {
+		return false
+	}
+
+	if b.State != Open {
+		return false
+	}
+
+	for _, s := range b.sweeps {
+		if s.IsExternalAddr || sweep.IsExternalAddr {
+			return false
+		}
+	}
+
+	for _, s := range b.sweeps {
+		timeoutDistance :=
+			int32(math.Abs(float64(sweep.Timeout - s.Timeout)))
+
+		if timeoutDistance > b.Cfg.MaxTimeoutDistance {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Wait waits for the batch to gracefully stop.
+func (b *Batch) Wait() {
+	b.log.Infof("Stopping")
+	b.wg.Wait()
+}
+
+// Run is the batch's main event loop.
+func (b *Batch) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+		b.wg.Wait()
+	}()
+
+	blockChan, blockErrChan, err :=
+		b.chainNotifier.RegisterBlockEpochNtfn(ctx)
+	if err != nil {
+		return err
+	}
+
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+
+		// Dispatch the spend monitor.
+		b.monitorSpend(runCtx)
+	}()
+
+	// We use a timer chan in order to trigger the batch publishment a few
+	// seconds after a new block is received. This is done avoid publishing
+	// the transaction on the same block in which it was spent. By waiting,
+	// we give the tiny amount of time needed by the spend monitor to
+	// receive the notification and close the batch, preventing further
+	// publishments.
+	var timerChan <-chan time.Time
+
+	b.log.Infof("started, primary %x, total sweeps %v",
+		b.PrimarySweepID[0:6], len(b.sweeps))
+
+	for {
+		select {
+		case sweep := <-b.RequestChan:
+			err := b.addSweep(runCtx, *sweep)
+			if err != nil {
+				return err
+			}
+
+		case height := <-blockChan:
+			b.log.Debugf("received block %v", height)
+
+			// We skip the step if this was the first height we
+			// received, as this means we just started running. This
+			// prevents immediately publishing a transaction on
+			// batch creation.
+			if b.currentHeight > 0 {
+				timerChan = time.After(batchPublishDelay)
+			} else {
+				b.log.Debugf("skipping step on first height")
+			}
+
+			b.currentHeight = height
+
+		case <-timerChan:
+			err := b.step(runCtx)
+			if err != nil {
+				return err
+			}
+
+		case spend := <-b.spendChan:
+			err := b.handleSpend(runCtx, spend.SpendingTx)
+			if err != nil {
+				return err
+			}
+
+		case <-b.confChan:
+			return b.handleConf(runCtx)
+
+		case err := <-blockErrChan:
+			return err
+
+		case err := <-b.errChan:
+			return err
+
+		case <-runCtx.Done():
+			return runCtx.Err()
+		}
+	}
+}
+
+// step executes the next step in the batch's lifecycle.
+func (b *Batch) step(ctx context.Context) error {
```

Comment:
I feel like this could be much improved in terms of readability which in return would also avoid the need for the tricks to handle the restart of the spend and confirmation monitors.

There are two cases when starting up a batch: 
1) it already has a primary sweep defined (read from the DB)
2) it's still empty => meaning no primary sweep

In case of 1) we'd start the spend monitor at the same time of starting `Run(...)`. So just check if the primary sweep's hash is non-zero and start the spend monitor (this is done currently too).

In the case of 2) we'd start the spend monitor when adding the first sweep.

There's really no need to restart the spend monitor on each publish.

For the conf monitor, we really only need to start it when we receive the spend notification. Since we only receive one spend notification we can start the conf montior right at the end of `handleSpend`. There's no need for starting it here in the `step` function. This way when we "step" we'd just attempt to publish. (I think we could completely remove the `step` function too.

Reply:
> 2) we'd start the spend monitor when adding the first sweep.

changed it so that each batch dispatches exactly one spend monitor over its runtime

> There's really no need to restart the spend monitor on each publish.

Yeap got rid of that too, same for conf monitor (it is now dispatched after detecting spend). Which also allowed getting rid of `step` altogether